src/ray/object_manager/push_manager.cc

```diff
@@ -29,81 +28,91 @@ void PushManager::StartPush(const NodeID &dest_id,
   // TODO(ekl) this isn't the best implementation of round robin, we should
   // consider tracking the number of chunks active per-push and balancing those.
+  // TODO(dayshah): Does round-robin even make sense here? We should probably finish
+  // pushes in the order they were asked for, so that some finish and some work can start.
+  // Otherwise all work will be halted for a period of time
+
+  // Loop over all active pushes for approximate round-robin prioritization.
+  bool keep_looping = true;
   while (chunks_in_flight_ < max_chunks_in_flight_ && keep_looping) {
     // Loop over each active push and try to send another chunk.
-    auto it = push_requests_with_chunks_to_send_.begin();
+    // If we could push out a chunk and haven't reached the chunks_in_flight_ limit,
+    // we'll loop again to try to send more chunks.
     keep_looping = false;
-    while (it != push_requests_with_chunks_to_send_.end() &&
+    auto iter = push_requests_with_chunks_to_send_.begin();
+    while (iter != push_requests_with_chunks_to_send_.end() &&
            chunks_in_flight_ < max_chunks_in_flight_) {
-      auto push_id = it->first;
-      auto &info = it->second;
-      if (info->SendOneChunk()) {
-        chunks_in_flight_ += 1;
-        keep_looping = true;
-        RAY_LOG(DEBUG) << "Sending chunk " << info->next_chunk_id << " of "
-                       << info->num_chunks << " for push " << push_id.first << ", "
-                       << push_id.second << ", chunks in flight " << NumChunksInFlight()
-                       << " / " << max_chunks_in_flight_
-                       << " max, remaining chunks: " << NumChunksRemaining();
-      }
-      if (info->NoChunksToSend()) {
-        it = push_requests_with_chunks_to_send_.erase(it);
+      auto &push_state = *iter;
+      push_state.SendOneChunk();
+      chunks_in_flight_ += 1;
+      if (push_state.num_chunks_to_send == 0) {
+        auto push_state_map_iter = push_state_map_.find(push_state.node_id);
+        RAY_CHECK(push_state_map_iter != push_state_map_.end());
+        auto &dest_map = push_state_map_iter->second;
+        auto dest_map_iter = dest_map.find(push_state.object_id);
+        RAY_CHECK(dest_map_iter != dest_map.end());
+        iter = push_requests_with_chunks_to_send_.erase(dest_map_iter->second);
+        dest_map.erase(dest_map_iter);
+        if (dest_map.empty()) {
+          push_state_map_.erase(push_state_map_iter);
+        }
       } else {
-        it++;
+        keep_looping = true;
+        iter++;
       }
     }
   }
 }
 
+void PushManager::HandleNodeRemoved(const NodeID &node_id) {
+  auto push_state_map_iter = push_state_map_.find(node_id);
+  if (push_state_map_iter == push_state_map_.end()) {
+    return;
+  }
+  for (auto &[_, push_state_iter] : push_state_map_iter->second) {
+    push_requests_with_chunks_to_send_.erase(push_state_iter);
+  }
+  push_state_map_.erase(node_id);
```

Comment:
I'd drop a comment here that erasing it from this map will cause the push loop to stop sending to this node